File: scala/0572-subtree-of-another-tree.scala
object Solution {
    def isSubtree(root: TreeNode, subRoot: TreeNode): Boolean = {
        if (root == null && subRoot == null) {
            return true
        } else if (root != null && subRoot == null) {
            return true
        } else if (root == null && subRoot != null) {
            return false
        } else {
            return isSameTree(root, subRoot) || (isSubtree(root.left, subRoot) || isSubtree(root.right, subRoot))
        }
    }
    
    def isSameTree(p: TreeNode, q: TreeNode): Boolean = {
        if (p == null && q == null) {
            return true
        } else if (p == null || q == null) {
            return false
        } else if (p.value != q.value) {
            return false 
        } else {
            return isSameTree(p.left, q.left) && isSameTree(p.right, q.right)
        }
    }
}

File: swift/0572-subtree-of-another-tree.swift
class Solution {
    private func areSameTrees(
        _ node1: TreeNode?, 
        _ node2: TreeNode?
    ) -> Bool {
        // if both are nil, they are same
        if node1 == nil && node2 == nil { return true }
        
        // if anyone else is nil, they are not the same
        guard let node1 = node1 else { return false }
        guard let node2 = node2 else { return false }
        
        return (
            (node1.val == node2.val) &&
            areSameTrees(node1.left, node2.left) &&
            areSameTrees(node1.right, node2.right)
        )
    }
    
    func isSubtree(_ root: TreeNode?, _ subRoot: TreeNode?) -> Bool {
        // Base cases (both can be nil, just any one can't)
        if root == nil && subRoot == nil { return true }
        guard let root = root else { return false }
        guard let subRoot = subRoot else { return false }
        
        // Base case: Exactly same tree
        if areSameTrees(root, subRoot) { return true }
        
        // Recursive
        return isSubtree(root.left, subRoot) 
                || isSubtree(root.right, subRoot)
    }
}